### Group Project 4 : Comparing 3 Models for Predicting Recidivism

For background on this project, please see the [README](../README.md).

**Notebooks**
- Data Acquisition & Cleaning (this notebook)
- [Exploratory Data Analysis](./02_eda.ipynb)
- [Modeling](./03_modeling.ipynb)
- [Results and Recommendations](./04_results.ipynb)

**In this notebook, you'll find (for each of the 3 models):**
- Data ingestion
- Cleaning
- New feature engineering
- etc. TODO

In [2]:
import pandas as pd

**Model 1: Base feature set - New York**

TODO provide some background

**Model 2: Criminal history feature set - Florida**

The Florida dataset comprises a number of SQLite tables related to criminal history for around 11,000 Broward County citizens. This data was collected as part of the evaluation of the Correctional Offender Management Profiling for Alternative Sanctions (COMPAS) decision support tool used in Broward County and in other U.S. states - further descriptions available on [Wikipedia](https://en.wikipedia.org/wiki/COMPAS_(software)) - so there are some COMPAS scores within this dataset as well.

In order to retrieve the data and put it into a form that was conducive to Pandas analysis, two SQL queries were constructed:

- [people.sql](../data/FL/people.sql) - basic demographic data, incarceration dates, and COMPAS scores
- [charges.sql](../data/FL/charges.sql) - case management, charge and arrest data

The results of those queries were then exported to CSV as [final_people.csv](../data/FL/final_people.csv) and [final_charges.csv](../data/fl/final_charges.csv) within the **data** subfolder, and will serve as the basis for the rest of this cleaning and transformation exercise.

In [104]:
# read in people dataset and take a look
people = pd.read_csv('../data/FL/final_people.csv')
people.head()

person_id   sex              race                  birth_date  \
0          1  Male             Other  1947-04-18 00:00:00.000000   
1          2  Male         Caucasian  1985-02-06 00:00:00.000000   
2          3  Male  African-American  1982-01-22 00:00:00.000000   
3          4  Male  African-American  1991-05-14 00:00:00.000000   
4          5  Male  African-American  1993-01-21 00:00:00.000000   

     first_incarceration_date first_incarceration_release  \
0  2013-08-13 06:03:42.000000  2013-08-14 05:41:20.000000   
1  2014-12-30 10:47:52.000000  2015-01-03 02:18:24.000000   
2  2015-10-15 00:00:00.000000  2015-12-07 00:00:00.000000   
3  2013-04-13 04:58:34.000000  2013-04-14 07:02:04.000000   
4                         NaN                         NaN   

      last_incarceration_date  last_incarceration_release  num_incarcerations  \
0  2014-07-07 09:26:12.000000  2014-07-14 08:24:15.000000                   2   
1  2014-12-30 10:47:52.000000  2015-01-03 02:18:24.000000                   1   
2  2013-01-26 03:45:27.000000  2013-02-05 05:36:53.000000                   2   
3  2016-01-08 09:59:55.000000  2016-01-09 04:41:39.000000                   5   
4                         NaN                         NaN                   0   

   comp_f_min_score  ...  comp_f_max_score  comp_f_max_decile  \
0                13  ...                13                  1   
1                16  ...                16                  2   
2                25  ...                25                  6   
3                26  ...                26                  7   
4                19  ...                19                  3   

   comp_r_min_score  comp_r_min_decile  comp_r_max_score  comp_r_max_decile  \
0             -2.78                  1             -2.78                  1   
1             -0.34                  5             -0.34                  5   
2             -0.76                  3             -0.76                  3   
3             -0.66                  4             -0.66                  4   
4              0.16                  8              0.16                  8   

   comp_v_min_score  comp_v_min_decile  comp_v_max_score  comp_v_max_decile  
0             -4.31                  1             -4.31                  1  
1             -2.75                  2             -2.75                  2  
2             -3.07                  1             -3.07                  1  
3             -2.26                  3             -2.26                  3  
4             -1.59                  6             -1.59                  6  

[5 rows x 21 columns]

In [105]:
# check people data types
people.dtypes

person_id                        int64
sex                             object
race                            object
birth_date                      object
first_incarceration_date        object
first_incarceration_release     object
last_incarceration_date         object
last_incarceration_release      object
num_incarcerations               int64
comp_f_min_score                 int64
comp_f_min_decile                int64
comp_f_max_score                 int64
comp_f_max_decile                int64
comp_r_min_score               float64
comp_r_min_decile                int64
comp_r_max_score               float64
comp_r_max_decile                int64
comp_v_min_score               float64
comp_v_min_decile                int64
comp_v_max_score               float64
comp_v_max_decile                int64
dtype: object

In [106]:
# convert all datetime columns
people['birth_date'] = pd.to_datetime(people['birth_date'])
people['first_incarceration_date'] = pd.to_datetime(people['first_incarceration_date'])
people['first_incarceration_release'] = pd.to_datetime(people['first_incarceration_release'])
people['last_incarceration_date'] = pd.to_datetime(people['last_incarceration_date'])
people['last_incarceration_release'] = pd.to_datetime(people['last_incarceration_release'])

In [107]:
# dummify sex and race
people = pd.get_dummies(data = people, columns = ['sex', 'race'], drop_first = True)

In [108]:
# final dtype/null check
people.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11757 entries, 0 to 11756
Data columns (total 25 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   person_id                    11757 non-null  int64         
 1   birth_date                   11757 non-null  datetime64[ns]
 2   first_incarceration_date     11127 non-null  datetime64[ns]
 3   first_incarceration_release  11127 non-null  datetime64[ns]
 4   last_incarceration_date      11127 non-null  datetime64[ns]
 5   last_incarceration_release   11127 non-null  datetime64[ns]
 6   num_incarcerations           11757 non-null  int64         
 7   comp_f_min_score             11757 non-null  int64         
 8   comp_f_min_decile            11757 non-null  int64         
 9   comp_f_max_score             11757 non-null  int64         
 10  comp_f_max_decile            11757 non-null  int64         
 11  comp_r_min_score             11757 non-nu

CONCLUSIONS
- Our data types are now looking good
- It appears that there are 630 people who have no incarceration data (the nulls from above) - they should be dropped from our analysis, since at least an initial incarceration is required to qualify for recidivism
- The COMPAS scores could be interesting for later comparison, but will not be used as features, so they can be saved to a separate file for now

In [109]:
people = people[people['num_incarcerations'] > 0]

In [110]:
compas = people[['person_id'] + [col for col in people if col.startswith('comp_')]]
compas.to_csv('../data/FL/final_compas.csv')
people = people.drop(columns = [col for col in people if col.startswith('comp_')])
people.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11127 entries, 0 to 11756
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   person_id                    11127 non-null  int64         
 1   birth_date                   11127 non-null  datetime64[ns]
 2   first_incarceration_date     11127 non-null  datetime64[ns]
 3   first_incarceration_release  11127 non-null  datetime64[ns]
 4   last_incarceration_date      11127 non-null  datetime64[ns]
 5   last_incarceration_release   11127 non-null  datetime64[ns]
 6   num_incarcerations           11127 non-null  int64         
 7   sex_Male                     11127 non-null  uint8         
 8   race_Asian                   11127 non-null  uint8         
 9   race_Caucasian               11127 non-null  uint8         
 10  race_Hispanic                11127 non-null  uint8         
 11  race_Native American         11127 non-nu

CONCLUSIONS
- Our demographic (birth date/sex/race) and incarceration statistics are now in appropriate format for feature engineering
- Now we need to incorporate data related to arrests and charges

In [111]:
# read in charges dataset and take a look
charges = pd.read_csv('../data/FL/final_charges.csv')
charges.head()

person_id    case_number                offense_date charge_degree  \
0          1  09083797TI30A  2009-08-11 00:00:00.000000           (0)   
1          1  09098832TI20A  2009-10-24 00:00:00.000000           (0)   
2          1  13009443TI30A  2013-01-14 00:00:00.000000           (0)   
3          1  13009443TI30A  2013-01-14 00:00:00.000000           (0)   
4          1  13011352CF10A  2013-08-13 00:00:00.000000          (F3)   

                             charge                 arrest_date  
0  Unlawful Speed (Requires Speeds)                         NaN  
1               Speed/65 Interstate                         NaN  
2             Disobey/Ran Stop Sign                         NaN  
3            Expired Tag/Infraction                         NaN  
4      Aggravated Assault w/Firearm  2013-08-13 00:00:00.000000

In [112]:
charges.dtypes

person_id         int64
case_number      object
offense_date     object
charge_degree    object
charge           object
arrest_date      object
dtype: object

In [113]:
# convert datetimes
charges['offense_date'] = pd.to_datetime(charges['offense_date'])
charges['arrest_date'] = pd.to_datetime(charges['arrest_date'])

In [114]:
# final dtypes/null check
charges.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140034 entries, 0 to 140033
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   person_id      140034 non-null  int64         
 1   case_number    140034 non-null  object        
 2   offense_date   140034 non-null  datetime64[ns]
 3   charge_degree  140034 non-null  object        
 4   charge         139605 non-null  object        
 5   arrest_date    92150 non-null   datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(3)
memory usage: 6.4+ MB


CONCLUSIONS
- We will collect first/last dates and charge degrees for charges and arrests per person as potentially useful features
- We will retrieve counts of charges by degree per person (pivoted) as well
- We will collect aggregated charge text data for NLP
- We will calculate mean time between offenses and between arrests

In [115]:
# Collect first/last charge and arrest data
# Using the offense date as our "first/last" marker
# And add to people dataframe
people['first_arrest_date'] = \
    people.join(charges.sort_values(by = ['person_id', 'offense_date']).groupby('person_id')['arrest_date'].first(),
    on = 'person_id',
    how = 'left')['arrest_date']
people['last_arrest_date'] = \
    people.join(charges.sort_values(by = ['person_id', 'offense_date']).groupby('person_id')['arrest_date'].last(),
    on = 'person_id',
    how = 'left')['arrest_date']
people['first_charge_degree'] = \
    people.join(charges.sort_values(by = ['person_id', 'offense_date']).groupby('person_id')['charge_degree'].first(),
    on = 'person_id',
    how = 'left')['charge_degree']
people['last_charge_degree'] = \
    people.join(charges.sort_values(by = ['person_id', 'offense_date']).groupby('person_id')['charge_degree'].last(),
    on = 'person_id',
    how = 'left')['charge_degree']

In [116]:
# pivot the degrees of charges into counts per person
# and add to people dataframe
charge_pivot = pd.pivot_table(charges, index = 'person_id', columns = 'charge_degree', aggfunc = 'count')['arrest_date']
charge_pivot.columns = ['charge_degree_count_' + col.replace('(', '').replace(')', '') for col in charge_pivot.columns]
people = people.join(charge_pivot, on = 'person_id', how = 'left')


In [117]:
# aggregate charges per person for NLP
# and add to people dataframe
# https://stackoverflow.com/questions/27298178/concatenate-strings-from-several-rows-using-pandas-groupby
agg_charges = charges[charges['charge'].notnull()].groupby('person_id').agg({'charge': '|'.join})
people['agg_charges'] = \
    people.join(agg_charges, on = 'person_id', how = 'left')['charge']

In [118]:
# calculate mean time between offenses and between arrests per person
# and add to people dataframe
# https://stackoverflow.com/questions/45241221/python-pandas-calculate-average-days-between-dates
charges['previous_offense'] = \
    charges.sort_values(by = ['person_id', 'offense_date']).groupby(['person_id'])['offense_date'].shift(1)
charges['days_between_offenses'] = \
    (charges['offense_date'] - charges['previous_offense']).apply(lambda x: x.days)
charges['previous_arrest'] = \
    charges.sort_values(by = ['person_id', 'arrest_date']).groupby(['person_id'])['arrest_date'].shift(1)
charges['days_between_arrests'] = \
    (charges['arrest_date'] - charges['previous_arrest']).apply(lambda x: x.days)

# need to remove zeros - don't want them affecting the averages
days_between_offenses = charges[charges['days_between_offenses'] > 0].groupby('person_id')['days_between_offenses'].mean()
days_between_arrests = charges[charges['days_between_arrests'] > 0].groupby('person_id')['days_between_arrests'].mean()

people['avg_days_between_offenses'] = \
    people.join(days_between_offenses, on = 'person_id', how = 'left')['days_between_offenses']
people['avg_days_between_arrests'] = \
    people.join(days_between_arrests, on = 'person_id', how = 'left')['days_between_arrests']

In [119]:
people.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11127 entries, 0 to 11756
Data columns (total 38 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   person_id                       11127 non-null  int64         
 1   birth_date                      11127 non-null  datetime64[ns]
 2   first_incarceration_date        11127 non-null  datetime64[ns]
 3   first_incarceration_release     11127 non-null  datetime64[ns]
 4   last_incarceration_date         11127 non-null  datetime64[ns]
 5   last_incarceration_release      11127 non-null  datetime64[ns]
 6   num_incarcerations              11127 non-null  int64         
 7   sex_Male                        11127 non-null  uint8         
 8   race_Asian                      11127 non-null  uint8         
 9   race_Caucasian                  11127 non-null  uint8         
 10  race_Hispanic                   11127 non-null  uint8         
 11  ra

CONCLUSIONS
- It appears that there are about 500 people with no arrest data - they should be dropped from our analysis
- The **first_charge_degree** and **last_charge_degree** columns should be dummified
- Zeros should be imputed for the **charge_degree_XXX** columns
- Empty string should be imputed for **agg_charges** column
- For people with nulls in the **avg_days_between_XXX** columns, who are probably non-recidivists, it makes sense to impute a very large number

In [120]:
people = people[people['first_arrest_date'].notnull()]
people.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10645 entries, 0 to 11756
Data columns (total 38 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   person_id                       10645 non-null  int64         
 1   birth_date                      10645 non-null  datetime64[ns]
 2   first_incarceration_date        10645 non-null  datetime64[ns]
 3   first_incarceration_release     10645 non-null  datetime64[ns]
 4   last_incarceration_date         10645 non-null  datetime64[ns]
 5   last_incarceration_release      10645 non-null  datetime64[ns]
 6   num_incarcerations              10645 non-null  int64         
 7   sex_Male                        10645 non-null  uint8         
 8   race_Asian                      10645 non-null  uint8         
 9   race_Caucasian                  10645 non-null  uint8         
 10  race_Hispanic                   10645 non-null  uint8         
 11  ra

In [121]:
people['first_charge_degree'] = people['first_charge_degree'].str.replace('(', '', regex = False).str.replace(')', '', regex = False)
people['last_charge_degree'] = people['last_charge_degree'].str.replace('(', '', regex = False).str.replace(')', '', regex = False)
people = pd.get_dummies(data = people, columns = ['first_charge_degree', 'last_charge_degree'], drop_first = True)

In [122]:
people[[col for col in people if col.startswith('charge_degree_')]] = people[[col for col in people if col.startswith('charge_degree_')]].fillna(value = 0)
people.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10645 entries, 0 to 11756
Data columns (total 63 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   person_id                       10645 non-null  int64         
 1   birth_date                      10645 non-null  datetime64[ns]
 2   first_incarceration_date        10645 non-null  datetime64[ns]
 3   first_incarceration_release     10645 non-null  datetime64[ns]
 4   last_incarceration_date         10645 non-null  datetime64[ns]
 5   last_incarceration_release      10645 non-null  datetime64[ns]
 6   num_incarcerations              10645 non-null  int64         
 7   sex_Male                        10645 non-null  uint8         
 8   race_Asian                      10645 non-null  uint8         
 9   race_Caucasian                  10645 non-null  uint8         
 10  race_Hispanic                   10645 non-null  uint8         
 11  ra

In [123]:
people[['agg_charges']] = people[['agg_charges']].fillna(value = '')
people.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10645 entries, 0 to 11756
Data columns (total 63 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   person_id                       10645 non-null  int64         
 1   birth_date                      10645 non-null  datetime64[ns]
 2   first_incarceration_date        10645 non-null  datetime64[ns]
 3   first_incarceration_release     10645 non-null  datetime64[ns]
 4   last_incarceration_date         10645 non-null  datetime64[ns]
 5   last_incarceration_release      10645 non-null  datetime64[ns]
 6   num_incarcerations              10645 non-null  int64         
 7   sex_Male                        10645 non-null  uint8         
 8   race_Asian                      10645 non-null  uint8         
 9   race_Caucasian                  10645 non-null  uint8         
 10  race_Hispanic                   10645 non-null  uint8         
 11  ra

In [124]:
people[['avg_days_between_offenses']] = people[['avg_days_between_offenses']].fillna(value = 999999)
people[['avg_days_between_arrests']] = people[['avg_days_between_arrests']].fillna(value = 999999)
people.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10645 entries, 0 to 11756
Data columns (total 63 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   person_id                       10645 non-null  int64         
 1   birth_date                      10645 non-null  datetime64[ns]
 2   first_incarceration_date        10645 non-null  datetime64[ns]
 3   first_incarceration_release     10645 non-null  datetime64[ns]
 4   last_incarceration_date         10645 non-null  datetime64[ns]
 5   last_incarceration_release      10645 non-null  datetime64[ns]
 6   num_incarcerations              10645 non-null  int64         
 7   sex_Male                        10645 non-null  uint8         
 8   race_Asian                      10645 non-null  uint8         
 9   race_Caucasian                  10645 non-null  uint8         
 10  race_Hispanic                   10645 non-null  uint8         
 11  ra

CONCLUSIONS
- We've resolved all null issues
- We lost about 1,000 rows, but the vast majority of those were people for whom recidivism was incalculable, so it's an acceptable loss
- We will keep agg_charges to the side for potential NLP analysis in vectorized form later
- We have 7 datetime columns that need to be reengineered into numerics, since we're not attempting time series analysis here - the most useful scenario is probably to convert the non-date-of-birth columns into "years since birth" quantities by referencing the date of birth column
- There are also some potentially interesting features we could engineer that summarize charges - total number of charges, number of felonies/misdemeanors/infractions

In [ ]:
# We'll use this to determine folks with no charges to drop, and potentially also useful as a feature
people['total_charge_count'] = people[[col for col in people if col.startswith('charge_degree_count_')]].sum(axis = 1)

**Model 3: Behavioral feature set - Georgia**

TODO provide some background

In [3]:
ga = pd.read_csv('../data/GA/NIJ_s_Recidivism_Challenge_Full_Dataset.csv')

# Lowercasing all column names and string values
ga.columns = [col.lower() for col in ga.columns]
ga = ga.applymap(lambda x: x.lower() if type(x) == str else x)

ga.head()

id gender   race age_at_release  residence_puma gang_affiliated  \
0   1      m  black          43-47              16           False   
1   2      m  black          33-37              16           False   
2   3      m  black    48 or older              24           False   
3   4      m  white          38-42              16           False   
4   5      m  white          33-37              16           False   

   supervision_risk_score_first supervision_level_first  \
0                           3.0                standard   
1                           6.0             specialized   
2                           7.0                    high   
3                           7.0                    high   
4                           4.0             specialized   

         education_level dependents  ... drugtests_meth_positive  \
0  at least some college  3 or more  ...                0.000000   
1   less than hs diploma          1  ...                0.000000   
2  at least some college  3 or more  ...                0.166667   
3   less than hs diploma          1  ...                0.000000   
4   less than hs diploma  3 or more  ...                0.058824   

  drugtests_other_positive percent_days_employed jobs_per_year  \
0                      0.0              0.488562      0.447610   
1                      0.0              0.425234      2.000000   
2                      0.0              0.000000      0.000000   
3                      0.0              1.000000      0.718996   
4                      0.0              0.203562      0.929389   

  employment_exempt recidivism_within_3years recidivism_arrest_year1  \
0             False                    False                   False   
1             False                     True                   False   
2             False                     True                   False   
3             False                    False                   False   
4             False                     True                    True   

  recidivism_arrest_year2  recidivism_arrest_year3  training_sample  
0                   False                    False                1  
1                   False                     True                1  
2                    True                    False                1  
3                   False                    False                1  
4                   False                    False                1  

[5 rows x 54 columns]

In [4]:
# Checking data types and null values
ga.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25835 entries, 0 to 25834
Data columns (total 54 columns):
 #   Column                                             Non-Null Count  Dtype  
---  ------                                             --------------  -----  
 0   id                                                 25835 non-null  int64  
 1   gender                                             25835 non-null  object 
 2   race                                               25835 non-null  object 
 3   age_at_release                                     25835 non-null  object 
 4   residence_puma                                     25835 non-null  int64  
 5   gang_affiliated                                    22668 non-null  object 
 6   supervision_risk_score_first                       25360 non-null  float64
 7   supervision_level_first                            24115 non-null  object 
 8   education_level                                    25835 non-null  object 
 9   depend

In [5]:
ga['race'].value_counts(normalize=True)

black    0.574686
white    0.425314
Name: race, dtype: float64

In [6]:
# Dropping unnecessary columns; explanations below
cols_to_drop = ['id', 'race', 'residence_puma', 'supervision_level_first', 'prison_offense', 'jobs_per_year', 
'prior_revocations_parole', 'prior_revocations_probation', 
'program_unexcusedabsences', 'employment_exempt',
'recidivism_arrest_year1', 'recidivism_arrest_year2', 'recidivism_arrest_year3',
'training_sample']

prior_conv_cols = [col for col in ga if col.startswith('prior_conv')]

old_shape = ga.shape
print(old_shape)
ga.drop(ga[cols_to_drop + prior_conv_cols], axis=1, inplace=True)
new_shape = ga.shape
print(new_shape)
print(f'Total columns dropped: {old_shape[1] - new_shape[1]}')

(25835, 54)
(25835, 32)
Total columns dropped: 22


- Dropped 'id' column because it is not relevant to our analysis, and we have confirmed that there are no repeated observations within this dataset.
- Because the probability of 'black' or 'white' being the only two options for race (shown above) in almost 26,000 observations seems very low, I'll be dropping this column rather than making any statements about race that cannot be confirmed in the data.
- 'Residence puma' indicates the [residence group](https://nij.ojp.gov/funding/nij-recidivism-forecasting-challenge.pdf#page=26&zoom=100,54,936) at prison release by the US Census Bureau Public Use Microdata Area [(PUMA)](http://proximityone.com/puma/puma2010ga.htm). We originally kept this column in the data for EDA, but because this column is nondescript and there are multiple counties included in each code, we did not find that these values provided meaningful information that could be used in our analysis.
- 'Supervision level first' refers to the first parole supervision level assignment, and is highly correlated with but less useful than the 'supervision risk score first' column.
- 'Prison offense' is defined as the primary prison conviction offense group, which is somewhat repetitive and less informative than the arrest and conviction columns.
- 'Jobs per year' is going to be highly correlated with 'percent days employed'.
- 'Prior revocations parole' and 'prior revocations probation' are meant to indicate an unsuccessful termination of supervision or a return to prison to serve the remaining sentence. Supervision is not something we are particularly interested in, and a return to prison would be indicated in the 'recidivism' target column.
- 'Program unexcused absences' is less informative than 'program attendances'.
- 'Employment exempt' will also be indicated in the other columns i.e. if 'employment exempt' == True, then 'percent days employed' will be 0.
- The three recidivism columns are being dropped because we are not looking at when the person was reincarated, but rather whether or not they were.
- 'Training sample' does not tell us anything about the observations.
- The 'prior conviction' columns are similar to the 'prior arrest' columns. Both are informative, but we are choosing to use only the prior arrest features in our modeling.

In [7]:
ga_null = ga.isnull().sum()
ga_null[ga_null > 0]

gang_affiliated                 3167
supervision_risk_score_first     475
avg_days_per_drugtest           6103
drugtests_thc_positive          5172
drugtests_cocaine_positive      5172
drugtests_meth_positive         5172
drugtests_other_positive        5172
percent_days_employed            462
dtype: int64

In [8]:
# Handling nulls

old_shape = ga.shape
print(old_shape)

# Nulls in 'avg_days_per_drugtest'
ga_drop = ga.loc[(ga['drugtests_thc_positive'].notna()) & (ga['avg_days_per_drugtest'].isna())]
drop_idx = [i for i in ga_drop.index]
ga.drop(axis=0, index=drop_idx, inplace=True)

# Nulls in the four drug test columns
ga[['avg_days_per_drugtest', 'drugtests_thc_positive', 'drugtests_cocaine_positive', 'drugtests_meth_positive', 'drugtests_other_positive']] = ga[['avg_days_per_drugtest', 'drugtests_thc_positive', 'drugtests_cocaine_positive', 'drugtests_meth_positive', 'drugtests_other_positive']].fillna(value=0)

# Nulls in the 'gang_affiliated' columns
ga['gang_affiliated'].fillna(False, inplace=True)

# Remaining nulls in the 'supervision_risk_score_first' and 'percent_days_employed' columns
ga.dropna(axis=0, inplace=True)

new_shape = ga.shape
print(new_shape)
print(f'Rows dropped: {old_shape[0] - new_shape[0]}')

(25835, 32)
(24426, 32)
Rows dropped: 1409


In [9]:
ga['gang_affiliated'].value_counts(normalize=True)

False    0.844592
True     0.155408
Name: gang_affiliated, dtype: float64

- Of the 6,103 nulls in the 'avg days per drugtest' column, 5,172 of these overlap with the nulls in the drug test columns. For this reason, we feel that it's reasonable to impute the overlapping values with 0s and remove the remaining 931 rows in the 'avg days per drugtest' column.
- Because ~83% of the 'gang affiliated' values are False, we are confident that in filling the 3,167 nulls with False we can expect to be correct here in at least 2,533 of these imputed observations.
- We feel confident that dropping the null 'supervision risk score first' and 'percent_days_employed' observations will not drastically affect our model.

In [10]:
# Selecting all object-type columns
object_type = ga.select_dtypes(include='object')

# Selecting specifically the object-type columns that contain the substring 'or more'
or_more_cols = [col for col in object_type.columns if object_type[col].str.contains('or more').any()]

# Replacing the substring with empty string
ga[or_more_cols] = ga[or_more_cols].replace({'or more': ''}, regex=True)

# Converting all values in these columns to integers
ga[or_more_cols] = ga[or_more_cols].apply(pd.to_numeric) 

- The reasoning behind the above conversions is that the scales were not equal (i.e. one column may have '3 or more' and another has '10 or more'). Further, we do not know what the 'or more' means; '6 or more' in the 'prior arrest episodes misd' column could mean 7 prior misdemeanor arrests or 50. This ambiguity gives us little to go on, whereas removing the 'or more' and converting the column values to integers allows us more flexibility in modeling while producing similarly-interpreted results.

In [11]:
# Selecting the non-numeric columns to clean and decrease the number of columns that will need to be dummified during modeling

ga.select_dtypes(exclude=['int64', 'float64']).head()

gender age_at_release  gang_affiliated        education_level  \
0      m          43-47            False  at least some college   
1      m          33-37            False   less than hs diploma   
2      m    48 or older            False  at least some college   
3      m          38-42            False   less than hs diploma   
4      m          33-37            False   less than hs diploma   

        prison_years  prior_arrest_episodes_dvcharges  \
0  more than 3 years                            False   
1  more than 3 years                             True   
2          1-2 years                             True   
3          1-2 years                            False   
4          1-2 years                             True   

   prior_arrest_episodes_guncharges  condition_mh_sa  condition_cog_ed  \
0                             False             True              True   
1                             False            False             False   
2                             False             True              True   
3                             False             True              True   
4                             False             True              True   

   condition_other  violations_electronicmonitoring  violations_instruction  \
0            False                            False                   False   
1            False                            False                    True   
2            False                            False                    True   
3            False                            False                   False   
4             True                            False                   False   

   violations_failtoreport  violations_movewithoutpermission  \
0                    False                             False   
1                    False                             False   
2                    False                              True   
3                    False                             False   
4                    False                             False   

   recidivism_within_3years  
0                     False  
1                      True  
2                      True  
3                     False  
4                      True

In [12]:
# Final cleaning steps

# Finding all true/false columns in the dataframe
bool_vars = [col for col in ga.columns if (False in ga[col].values) and (True in ga[col].values)]
true_false = ga[bool_vars].select_dtypes(include='bool')

# Converting the values to numbers
for col in true_false.columns:
    if col in ga.columns:
        ga[col] = ga[col].map({True: 1, False: 0})

# Mapping gender to numerical values 
ga.rename(columns={'gender': 'gender_male'}, inplace=True) 
ga['gender_male'] = ga['gender_male'].map({'m': 1, 'f': 0}) 

# Ensuring that the changes were made
ga.head()

gender_male age_at_release  gang_affiliated  supervision_risk_score_first  \
0            1          43-47                0                           3.0   
1            1          33-37                0                           6.0   
2            1    48 or older                0                           7.0   
3            1          38-42                0                           7.0   
4            1          33-37                0                           4.0   

         education_level  dependents       prison_years  \
0  at least some college           3  more than 3 years   
1   less than hs diploma           1  more than 3 years   
2  at least some college           3          1-2 years   
3   less than hs diploma           1          1-2 years   
4   less than hs diploma           3          1-2 years   

   prior_arrest_episodes_felony  prior_arrest_episodes_misd  \
0                             6                           6   
1                             7                           6   
2                             6                           6   
3                             8                           6   
4                             4                           4   

   prior_arrest_episodes_violent  ...  delinquency_reports  \
0                              1  ...                    0   
1                              3  ...                    4   
2                              3  ...                    4   
3                              0  ...                    0   
4                              3  ...                    0   

   program_attendances  residence_changes  avg_days_per_drugtest  \
0                    6                  2             612.000000   
1                    0                  2              35.666667   
2                    6                  0              93.666667   
3                    6                  3              25.400000   
4                    7                  0              23.117647   

   drugtests_thc_positive  drugtests_cocaine_positive  \
0                0.000000                         0.0   
1                0.000000                         0.0   
2                0.333333                         0.0   
3                0.000000                         0.0   
4                0.000000                         0.0   

   drugtests_meth_positive  drugtests_other_positive  percent_days_employed  \
0                 0.000000                       0.0               0.488562   
1                 0.000000                       0.0               0.425234   
2                 0.166667                       0.0               0.000000   
3                 0.000000                       0.0               1.000000   
4                 0.058824                       0.0               0.203562   

   recidivism_within_3years  
0                         0  
1                         1  
2                         1  
3                         0  
4                         1  

[5 rows x 32 columns]

In [13]:
ga.to_csv('../data/GA/georgia_cleaned.csv')

**FINAL NOTES**:
- The final datasets for modeling are exported:
  - [here](../data/NY/NY_final.csv) for Model 1 (NY)
  - [here](../data/FL/FL_final.csv) for Model 2 (FL)
  - [here](../data/GA/GA_final.csv) for Model 3 (GA)
- The next notebook in the series is [Exploratory Data Analysis](./02_eda.ipynb).